File: Using_Java/LinkedLists/SinglyCircularLL.java
class NodeS
{
	public int iData;
	public NodeS nextPtr;
	
	public NodeS()
	{
		this.iData 	= 0;
		this.nextPtr 	= null;
	}
	
	public NodeS(int iValue)
	{
		this.iData   	= iValue;
		this.nextPtr 	= null;
	}
	
}

interface LinkedList
{
	public void insertFirst(int iNo);
	public void insertLast(int iNo);
	
	public void deleteFirst();
	public void deleteLast();
	
	public void insertAtPos(int iNo,int iPos);
	public void deleteAtPos(int iPos);
}

class SinglyCircular implements LinkedList
{
	private int iCount;	
	
	private NodeS first;
	private NodeS last;
	
	public SinglyCircular()
	{
		iCount	= 0;
		first 	= null;
		last	= null;
	} 
	
	public void displayNodes()
	{
		int iCnt = 0;
		NodeS tempPtr = first;
		
		for(iCnt = 0; iCnt < iCount; iCnt++)
		{
			System.out.print("|"+ tempPtr.iData+"|<->");
			tempPtr = tempPtr.nextPtr;
		}
		
	}
	
	public int countNodes()
	{
		return iCount;
	}
	
	public void insertFirst(int iNo)
	{
		NodeS newNode = null;
		
		newNode = new NodeS(iNo);
		
		if((first == null) && (last == null))
		{
			first = newNode;
			last  = newNode;
		}
		else
		{
			newNode.nextPtr = first;
			first = newNode;
		}
		
		last.nextPtr = first;
		
		iCount++;
	
	}
	
	public void insertLast(int iNo)
	{
		NodeS newNode = null;
		
		newNode = new NodeS(iNo);
		
		if((first == null) && (last == null))
		{
			first = newNode;
			last  = newNode;
		}
		else
		{
			last.nextPtr = newNode;
			last = last.nextPtr;
		}
		
		last.nextPtr = first;
		
		iCount++;
	}
	
	public void deleteFirst()
	{
		if((first == null) && (last == null))
		{
			System.out.println("\nLinked list is empty");
			return;
		}
		
		if(first == last)
		{
			first = null;
			last  = null;
		}
		else
		{
			first = first.nextPtr;
			last.nextPtr = first;
		}
		
		iCount--;
	}
	
	public void deleteLast()
	{
		if((first == null) && (last == null))
		{
			System.out.println("\nLinked list is empty");
			return;
		}
		
		if(first == last)
		{
			first = null;
			last  = null;
		}
		else
		{
			NodeS tempPtr = first;
			
			while(tempPtr.nextPtr != last)
			{
				tempPtr = tempPtr.nextPtr;
			}
			
			last = tempPtr;
			last.nextPtr = first;
		}
		
		iCount--;
	
	}
	
	public void insertAtPos(int iNo,int iPos)
	{
		if((iPos < 0) || (iPos > iCount + 1))
		{
			System.out.println("Invalid position\n");
			return;
		}
		
		if(iPos == 1)
		{
			insertFirst(iNo);
		}
		else if(iPos == iCount + 1)
		{
			insertLast(iNo);
		}
		else
		{
			int iCnt = 0;
			NodeS tempPtr = first;
			
			for(iCnt = 1; iCnt < iPos - 1; iCnt++)
			{
				tempPtr = tempPtr.nextPtr;
			}
			
			NodeS newNode = null;			
			newNode = new NodeS(iNo);
			
			newNode.nextPtr = tempPtr.nextPtr;
			tempPtr.nextPtr = newNode;
			
			iCount++;
		}
	
	}
	
	public void deleteAtPos(int iPos)
	{
		if((iPos < 0) || (iPos > iCount + 1))
		{
			System.out.println("Invalid position\n");
			return;
		}
		
		if(iPos == 1)
		{
			deleteFirst();
		}
		else if(iPos == iCount)
		{
			deleteLast();
		}
		else
		{
			int iCnt = 1;
			NodeS tempPtr  = first;
			NodeS tempPtrX = null; 
			
			for(iCnt = 1; iCnt < iPos - 1; iCnt++)
			{
				tempPtr = tempPtr.nextPtr;
			}
			
			tempPtrX = tempPtr.nextPtr;
			tempPtr.nextPtr = tempPtr.nextPtr.nextPtr;
			
			iCount--;
		}
		
	}

}

class SinglyCircularLL
{
	public static void main(java.lang.String Arg[])
	{
		SinglyCircular sObj = new SinglyCircular();
		
	System.out.println("============================================================================");
	
	
	System.out.println("\n----------------------------------------------------------------------------");
		System.out.println("Linked list after inserting nodes at first position : ");
		System.out.println("----------------------------------------------------------------------------");
	
		sObj.insertFirst(101);
		sObj.insertFirst(51);
		sObj.insertFirst(21);
		sObj.insertFirst(11);
		sObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+sObj.countNodes());


	System.out.println("\n----------------------------------------------------------------------------");	
		sObj.insertLast(111);
		sObj.insertLast(151);

		System.out.println("Linked list after inserting nodes at last position : ");
		System.out.println("----------------------------------------------------------------------------");		
		sObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+sObj.countNodes());

	System.out.println("\n----------------------------------------------------------------------------");	
		sObj.insertAtPos(55,4);

		System.out.println("Linked list after inserting node at given position : ");
		System.out.println("----------------------------------------------------------------------------");		
		sObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+sObj.countNodes());

	System.out.println("\n----------------------------------------------------------------------------");	
		sObj.deleteAtPos(4);

		System.out.println("Linked list after deleting node at given position : ");
		System.out.println("----------------------------------------------------------------------------");		
		sObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+sObj.countNodes());


	System.out.println("\n----------------------------------------------------------------------------");	
		
		sObj.deleteFirst();		
		
		System.out.println("Linked list after deleting node from first position : ");
		System.out.println("----------------------------------------------------------------------------");

		sObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+sObj.countNodes());

System.out.println("\n----------------------------------------------------------------------------");	
		
		sObj.deleteLast();
		System.out.println("Linked list after deleting node from last position : ");
		System.out.println("----------------------------------------------------------------------------");

		sObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+sObj.countNodes());


	System.out.println("\n============================================================================");
			
	}
	
}

File: Using_Python/LinkedLists/SinglyCircularLL.py
from abc import ABC,abstractmethod

class NodeS:
	
	def __init__(self,iElement):
		
		self.iData 	= iElement
		self.nextPtr 	= None

class LinkedLists(ABC):
	
	@abstractmethod
	def insertFirst(self,iNo):
		pass
			
	@abstractmethod
	def insertLast(self,iNo):
		pass
	
	@abstractmethod
	def deleteFirst(self):
		pass

	@abstractmethod		
	def deleteLast(self):
		pass
	
	@abstractmethod
	def insertAtPos(self,iNo,iPos):
		pass
	
	@abstractmethod	
	def deleteAtPos(self,iPos):
		pass
	

class SinglyCircularLL(LinkedLists):
	
	def __init__(self):
		self.iCount	= 0
		self.first	= None
		self.last	= None	 		
	
	def displayNodes(self):
		
		tempPtr = self.first	
		
		for iCnt in range(0,self.iCount,1):
			
			print("|",tempPtr.iData,"|<->",end = ' ')
			tempPtr = tempPtr.nextPtr
		print()
			
	def countNodes(self):
		
		return self.iCount
	
	def insertFirst(self,iValue):	
		
		newNode = None
		
		newNode = NodeS(iValue)
		
		if((self.first == None) and (self.last == None)):
			
			self.first = newNode
			self.last  = newNode
		
		else:
			
			newNode.nextPtr = self.first
			self.first = newNode
			
		
		self.last.nextPtr = self.first
		
		self.iCount = self.iCount + 1	
				
	
	def insertLast(self,iValue):
	
		newNode = None
		
		newNode = NodeS(iValue)
		
		if((self.first == None) and (self.last == None)):
			
			self.first = newNode
			self.last  = newNode
		
		else:
			self.last.nextPtr = newNode
			self.last = self.last.nextPtr		
		
		self.last.nextPtr = self.first
		
		self.iCount = self.iCount + 1	
				
		
	def deleteFirst(self):
		
		if(self.iCount == 0):
			
			print("Linked list is empty")
			return
		
		elif(self.iCount == 1):
			
			self.first = None
			self.last  = None
		
		else:
			
			self.first = self.first.nextPtr
			
			self.last.nextPtr = self.first
		
		self.iCount = self.iCount - 1			
			
	def deleteLast(self):

		if(self.iCount == 0):
			
			print("Linked list is empty")
			return
		
		elif(self.iCount == 1):
			
			self.first = None
			self.last  = None
		
		else:
			
			tempPtr = self.first
			
			while(tempPtr.nextPtr != self.last):
				
				tempPtr = tempPtr.nextPtr
				
			self.last = tempPtr
			
			self.nextPtr = self.first		
					
		self.iCount = self.iCount - 1			
	
	def insertAtPos(self,iValue,iPos):
		
		if((iPos < 1) or (iPos > self.iCount + 1)):	
			
			print("Invalid position")
			
		if(iPos == 1):
			
			insertFirst(iValue)
			
		elif(iPos == self.iCount + 1):
			
			insertLast(iValue)
			
		else:
			
			tempPtr = self.first
			
			for iCnt in range(1,iPos - 1,1):
				
				tempPtr = tempPtr.nextPtr
			
			newNode = None
			
			newNode = NodeS(iValue)	
			
			newNode.nextPtr = tempPtr.nextPtr
			tempPtr.nextPtr = newNode
			
			self.iCount = self.iCount + 1
			
	
	def deleteAtPos(self,iPos):
	
		if((iPos < 1) or (iPos > self.iCount)):	
			
			print("Invalid position")
			
		if(iPos == 1):
			
			deleteFirst()
			
		elif(iPos == self.iCount):
			
			deleteLast()
			
		else:
			
			tempPtr = self.first
			tempPtrX = None
			
			for iCnt in range(1,iPos - 1,1):
				
				tempPtr = tempPtr.nextPtr
			
			tempPtrX = tempPtr.nextPtr
				
			tempPtr.nextPtr = tempPtr.nextPtr.nextPtr
			
			self.iCount = self.iCount - 1
			
 	
def main():
	
	line = "-"*70
	
	print()
	print(line)
	print("Implementation of singly circular linked list using python")
	print(line)
	
	sObj = SinglyCircularLL()
	
	print("\nInserting elements at first position :")
	sObj.insertFirst(111)
	sObj.insertFirst(101)
	sObj.insertFirst(51)
	sObj.insertFirst(21)
	sObj.insertFirst(11)
	
	sObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(sObj.countNodes()))
	
	print("\nInserting node at given position : ")
	sObj.insertAtPos(55,4)
	
	sObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(sObj.countNodes()))
	
	print("\nDeleting node from given position : ")
	sObj.deleteAtPos(4)

	sObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(sObj.countNodes()))
	
	print("\nInserting elements at last position : ")
	sObj.insertLast(151)
	sObj.insertLast(201)
	sObj.insertLast(251)
	
	sObj.displayNodes()	
	print("Number of nodes in linked lists are : {}".format(sObj.countNodes()))
	
	print("\nDeleting node from first position : ")
	sObj.deleteFirst()
	
	sObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(sObj.countNodes()))
	
	print("\nDeleting node from last position : ")
	sObj.deleteLast()
	
	sObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(sObj.countNodes()))
	
	
if __name__ == "__main__":
	main()
